hal/src/dma.rs

```diff
@@ -26,721 +26,1050 @@
-    // this "does not play well with others" as it assumes _channelMask
-    // is the final arbiter of channels in use (this is true only within
-    // this library -- but other DMA-driven code may have allocated its
-    // own channel(s) elsewhere, sometimes with an equally broken
-    // approach).  A possible alternate approach, I haven't tested this
-    // yet, might be to loop through each channel, set DMAC->CHID.bit.ID
-    // and then test whether CHCTRLA.bit.ENABLE is set?  But for now...
-    for(channel=0; (channel < DMAC_CH_NUM) &&
-      (_channelMask & (1 << channel)); channel++);
-    // Doesn't help that code later does a software reset of the DMA
-    // controller, which would blow out other DMA-using libraries
-    // anyway (or they're just as likely to blow out this one).
-    // I think it's just an all-or-nothing affair...use one library
-    // for DMA everything, never mix and match.
-
-    if(channel >= DMAC_CH_NUM) // No free channel!
-        return DMA_STATUS_ERR_NOT_FOUND;
-
-    cpu_irq_enter_critical();
-
-    if(!_channelMask) { // No channels allocated yet; initialize DMA!
-#if !defined(DMAC_RESERVED_CHANNELS)
-#if (SAML21) || (SAML22) || (SAMC20) || (SAMC21)
-        PM->AHBMASK.bit.DMAC_       = 1;
-#elif defined(__SAMD51__)
-        MCLK->AHBMASK.bit.DMAC_     = 1; // Initialize DMA clocks
-#else
-        PM->AHBMASK.bit.DMAC_       = 1; // Initialize DMA clocks
-        PM->APBBMASK.bit.DMAC_      = 1;
-#endif
-        DMAC->CTRL.bit.DMAENABLE    = 0; // Disable DMA controller
-        DMAC->CTRL.bit.SWRST        = 1; // Perform software reset
-
-        // Initialize descriptor list addresses
-        DMAC->BASEADDR.bit.BASEADDR = (uint32_t)_descriptor;
-        DMAC->WRBADDR.bit.WRBADDR   = (uint32_t)_writeback;
-        memset(_descriptor, 0, sizeof(_descriptor));
-        memset(_writeback , 0, sizeof(_writeback));
-
-        // Re-enable DMA controller with all priority levels
-        DMAC->CTRL.reg = DMAC_CTRL_DMAENABLE | DMAC_CTRL_LVLEN(0xF);
-#endif
-
-        // Enable DMA interrupt at lowest priority
-#ifdef __SAMD51__
-        IRQn_Type irqs[] = { DMAC_0_IRQn, DMAC_1_IRQn, DMAC_2_IRQn,
-          DMAC_3_IRQn, DMAC_4_IRQn };
-        for(uint8_t i=0; i<(sizeof irqs / sizeof irqs[0]); i++) {
-            NVIC_EnableIRQ(irqs[i]);
-            NVIC_SetPriority(irqs[i], (1<<__NVIC_PRIO_BITS)-1);
-        }
-#else
-        NVIC_EnableIRQ(DMAC_IRQn);
-        NVIC_SetPriority(DMAC_IRQn, (1 << __NVIC_PRIO_BITS) - 1);
-#endif
+    // DMA CHANNEL FUNCTIONS ---------------------------------------------------
+
+    /// Allocates channel for DMA object
+    pub fn allocate(
+        &mut self,
+        clock_source: &mut ClockSource,
+        int_state: &InterruptState,
+    ) -> Status {
+        if self.channel.is_some() {
+            return Status::Ok; // Already alloc'd!
+        }
+
+        // Find index of first free DMA channel.  As currently written,
```

Comment:
This part of automatically finding a free channel felt a bit adafruit-y to me. Maybe for more control we should let the user figure it out? Idk, see what other rust dma implementations are doing.

Reply:
Changed to take an explicit `channel_id`, and return `Err(Error::Busy)` if it's in use